File: source/com/barrybecker4/simulation/liquid/LiquidSimulator.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.liquid;

import com.barrybecker4.common.util.FileUtil;
import com.barrybecker4.optimization.Optimizer;
import com.barrybecker4.optimization.parameter.NumericParameterArray;
import com.barrybecker4.optimization.parameter.ParameterArray;
import com.barrybecker4.optimization.parameter.types.Parameter;
import com.barrybecker4.optimization.strategy.OptimizationStrategyType;
import com.barrybecker4.simulation.common.ui.Simulator;
import com.barrybecker4.simulation.common.ui.SimulatorOptionsDialog;
import com.barrybecker4.simulation.liquid.config.ConfigurationEnum;
import com.barrybecker4.simulation.liquid.model.LiquidEnvironment;
import com.barrybecker4.simulation.liquid.rendering.EnvironmentRenderer;
import com.barrybecker4.simulation.liquid.rendering.RenderingOptions;
import com.barrybecker4.ui.util.GUIUtil;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;

/**
 * Main class for particle liquid simulation.
 *
 * @author Barry Becker
 */
public class LiquidSimulator extends Simulator implements MouseListener {

    private LiquidEnvironment environment_;
    private EnvironmentRenderer envRenderer_;

    /** These options can be changed while the simulation is running. */
    private LiquidDynamicOptions dynamicOptions_;

    /** The initial time step. It may adapt. */
    private static final double INITIAL_TIME_STEP = 0.005;

    private static final Color BG_COLOR = Color.white;

    private static final int NUM_OPT_PARAMS = 3;

    private boolean advectionOnly = false;

    /**
     * Constructor
     */
    public LiquidSimulator() {
        super("Liquid");

        environment_ = new LiquidEnvironment( ConfigurationEnum.getDefaultValue().getFileName());
        commonInit();
    }

    public void loadEnvironment(String configFile) {
        environment_ = new LiquidEnvironment(configFile);
        environment_.setAdvectionOnly(advectionOnly);
        commonInit();
    }

    @Override
    protected void reset() {
        boolean oldPaused = this.isPaused();
        setPaused(true);
        environment_.reset();
        commonInit();
        setPaused(oldPaused);
    }

    private void commonInit() {
        initCommonUI();
        envRenderer_ = new EnvironmentRenderer(environment_);

        int s = (int) envRenderer_.getScale();
        setPreferredSize(new Dimension( environment_.getWidth() * s, environment_.getHeight() * s));
    }

    @Override
    protected SimulatorOptionsDialog createOptionsDialog() {
         return new LiquidOptionsDialog( frame_, this );
    }

    @Override
    protected double getInitialTimeStep() {
        return INITIAL_TIME_STEP;
    }

    /**
     * @return a new recommended time step change.
     */
    @Override
    public double timeStep() {

        if ( !isPaused() ) {
            timeStep_ = environment_.stepForward( timeStep_);
        }
        return timeStep_;
    }

    public LiquidEnvironment getEnvironment() {
        return environment_;
    }

    @Override
    public void setScale( double scale ) {
        envRenderer_.setScale(scale);
    }
    @Override
    public double getScale() {
        return envRenderer_.getScale();
    }

    public RenderingOptions getRenderingOptions() {
        return envRenderer_.getRenderingOptions();
    }

    public boolean getSingleStepMode() {
        return !isAnimating();
    }

    public void setSingleStepMode(boolean singleStep) {
        setAnimating(!singleStep);
        if (singleStep)  {
            addMouseListener(this);
        }
        else {
            removeMouseListener(this);
        }
    }

    public boolean getAdvectionOnly() {
        return advectionOnly;
    }
    public void setAdvectionOnly(boolean advectOnly) {
        advectionOnly = advectOnly;
        environment_.setAdvectionOnly(advectOnly);
    }

    @Override
    public JPanel createDynamicControls() {
        dynamicOptions_ = new LiquidDynamicOptions(this);
        return dynamicOptions_;
    }

    @Override
    public void doOptimization() {

        Optimizer optimizer;
        if (GUIUtil.hasBasicService())
            optimizer = new Optimizer( this );
        else
            optimizer = new Optimizer( this, FileUtil.getHomeDir()+ "performance/liquid/liquid_optimization.txt" );
        Parameter[] params = new Parameter[3];
        ParameterArray paramArray = new NumericParameterArray( params );

        setPaused(false);
        optimizer.doOptimization(OptimizationStrategyType.GENETIC_SEARCH, paramArray, 0.3);
    }

    @Override
    public Color getBackground()  {
        return BG_COLOR;
    }

    @Override
    public void paint( Graphics g ) {
        if (g==null) return;
        Graphics2D g2 = (Graphics2D) g;
        envRenderer_.render(g2, getWidth(),  getHeight());
    }

    @Override
    protected String getFileNameBase() {
        return FileUtil.getHomeDir() + "temp/animations/simulation/" + "liquid/liquidFrame";
    }

    public void mouseClicked(MouseEvent e) {
        //System.out.println("mclick timeStep="+ timeStep_ );
        environment_.stepForward( timeStep_);
        this.repaint();
    }

    public void mousePressed(MouseEvent e) {}
    public void mouseReleased(MouseEvent e) {}
    public void mouseEntered(MouseEvent e) {}
    public void mouseExited(MouseEvent e) {}
}

File: scala-source/com/barrybecker4/simulation/liquid/LiquidSimulator.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.liquid

import com.barrybecker4.common.util.FileUtil
import com.barrybecker4.simulation.common.ui.Simulator
import com.barrybecker4.simulation.liquid.LiquidSimulator.{INITIAL_HEIGHT, INITIAL_WIDTH}
import com.barrybecker4.simulation.liquid.config.ConfigurationEnum
import com.barrybecker4.simulation.liquid.mpm.{MpmEnvironment, WaterEnvironment}
import com.barrybecker4.simulation.liquid.rendering.{EnvironmentRenderer, RenderingOptions}

import javax.swing.*
import java.awt.*
import scala.compiletime.uninitialized


/**
  * Main class for particle liquid simulation.
  * @author Barry Becker
  */
object LiquidSimulator {
  private val BG_COLOR = Color.gray
  private val INITIAL_WIDTH = 900
  private val INITIAL_HEIGHT = 900
}

class LiquidSimulator extends Simulator("Liquid") {

  private var environment: MpmEnvironment = new WaterEnvironment(ConfigurationEnum.DEFAULT_VALUE.fileName)
  private var envRenderer: EnvironmentRenderer = uninitialized
  /** These options can be changed while the simulation is running. */
  private var dynamicOptions: LiquidDynamicOptions = uninitialized
  commonInit()

  private[liquid] def loadEnvironment(configFile: String): Unit = {
    environment = new WaterEnvironment(configFile)
    commonInit()
  }

  override protected def reset(): Unit = {
    val oldPaused = this.isPaused
    setPaused(true)
    environment.reset()
    commonInit()
    setPaused(oldPaused)
  }

  private def commonInit(): Unit = {
    initCommonUI()
    envRenderer = new EnvironmentRenderer(environment)

    // Add the renderer to this component to enable proper event handling
    setLayout(new BorderLayout())
    removeAll() // Clear any existing components
    add(envRenderer, BorderLayout.CENTER)

    setPreferredSize(new Dimension(environment.getWidth, environment.getHeight))
    environment.initialize()
  }

  override protected def createOptionsDialog = new LiquidOptionsDialog(frame, this)
  override protected def getInitialTimeStep: Double = MpmEnvironment.DEFAULT_TIMESTEP
  def getEnvironment: MpmEnvironment = environment
  def getRenderingOptions: RenderingOptions = envRenderer.getRenderingOptions
  override def getBackground: Color = LiquidSimulator.BG_COLOR
  override protected def getFileNameBase: String = FileUtil.getHomeDir + "temp/animations/simulation/liquid/liquidFrame"

  /** @return a new recommended time step change. */
  override def timeStep: Double = {
    if (!isPaused) tStep = environment.advance()
    tStep
  }
  
  override def createDynamicControls: JPanel = {
    dynamicOptions = new LiquidDynamicOptions(this)
    dynamicOptions
  }

  override def doOptimization(): Unit = {
    throw new UnsupportedOperationException("Optimization not supported in liquid simulator.")
  }

  override def paint(g: Graphics): Unit = {
    val g2 = g.asInstanceOf[Graphics2D]
    envRenderer.render(g2, getWidth, getHeight)
  }
}
